trantor/net/EventLoop.cc

```diff
@@ -170,44 +170,96 @@ void EventLoop::quit()
+{
+template <typename F>
+struct ScopeExit
+{
+    ScopeExit(F &&f) : m_f(std::forward<F>(f))
+    {
+    }
+    ~ScopeExit()
+    {
+        m_f();
+    }
+    F m_f;
+};
+
+template <typename F>
+ScopeExit<F> makeScopeExit(F &&f)
+{
+    return ScopeExit<F>(std::forward<F>(f));
+};
+}  // namespace
+
 void EventLoop::loop()
 {
     assert(!looping_);
     assertInLoopThread();
     looping_.store(true, std::memory_order_release);
     quit_.store(false, std::memory_order_release);
 
-    while (!quit_.load(std::memory_order_acquire))
-    {
-        activeChannels_.clear();
+    std::exception_ptr loopException;
+    try
+    {  // Scope where the loop flag is set
+
+        auto loopFlagCleaner = makeScopeExit(
+            [this]() { looping_.store(false, std::memory_order_release); });
+        while (!quit_.load(std::memory_order_acquire))
+        {
+            activeChannels_.clear();
 #ifdef __linux__
-        poller_->poll(kPollTimeMs, &activeChannels_);
+            poller_->poll(kPollTimeMs, &activeChannels_);
 #else
-        poller_->poll(static_cast<int>(timerQueue_->getTimeout()),
-                      &activeChannels_);
-        timerQueue_->processTimers();
+            poller_->poll(static_cast<int>(timerQueue_->getTimeout()),
+                          &activeChannels_);
+            timerQueue_->processTimers();
 #endif
-        // TODO sort channel by priority
-        // std::cout<<"after ->poll()"<<std::endl;
-        eventHandling_ = true;
-        for (auto it = activeChannels_.begin(); it != activeChannels_.end();
-             ++it)
-        {
-            currentActiveChannel_ = *it;
-            currentActiveChannel_->handleEvent();
+            // TODO sort channel by priority
+            // std::cout<<"after ->poll()"<<std::endl;
+            eventHandling_ = true;
+            for (auto it = activeChannels_.begin(); it != activeChannels_.end();
+                 ++it)
+            {
+                currentActiveChannel_ = *it;
+                currentActiveChannel_->handleEvent();
+            }
+            currentActiveChannel_ = nullptr;
+            eventHandling_ = false;
+            // std::cout << "looping" << endl;
+            doRunInLoopFuncs();
         }
-        currentActiveChannel_ = nullptr;
-        eventHandling_ = false;
-        // std::cout << "looping" << endl;
-        doRunInLoopFuncs();
+        // loopFlagCleaner clears the loop flag here
+    }
+    catch (std::exception &e)
+    {
+        LOG_WARN << "Exception thrown from event loop, rethrowing after "
+                    "running functions on quit";
+        loopException = std::current_exception();
     }
-    looping_.store(false, std::memory_order_release);
 
+    // Run the quit functions even if exceptions were thrown
+    // TOOD: if more exceptions are thrown in the quit functions, some are left
+    // un-run. Can this be made exception safe?
     Func f;
     while (funcsOnQuit_.dequeue(f))
     {
         f();
     }
```

Comment:
Minor type. TODO

I think this is fine. Otherwise we run into the problem of needing to handle exceptions recursively. That sounds like a loosing game to me.

Reply:
Fixed typo.

That's a good point about recursive exceptions. C++ itself refuses to do that by forbidding exceptions from destructors when already unwinding an exception.

Which actually makes me think that running these quit functions in the scope exit would be the right alternative: as it is currently stands, if one of the exit functions throws, the original  exception gets lost (except for the log entry that there was an exception). OTOH, if we ran them on the scope exit and an exit function throws, we get terminate, which could be better (though at least in GCC, any info about the previous  exception gets lost anyway). Anyway, I'm leaving this as is for now, and maybe follow  up with another PR later to move them to the scope exit, if you think that would be the right thing?